File: app/src/main/java/com/sjn/stamp/ui/activity/MediaBrowserActivity.java
package com.sjn.stamp.ui.activity;

import android.content.ComponentName;
import android.os.Bundle;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import com.sjn.stamp.MusicService;
import com.sjn.stamp.ui.MediaBrowsable;
import com.sjn.stamp.ui.observer.MediaControllerObserver;
import com.sjn.stamp.utils.DrawerHelper;
import com.sjn.stamp.utils.LogHelper;

/**
 * Base activity for activities that need to show a playback control fragment when media is playing.
 */
public abstract class MediaBrowserActivity extends DrawerActivity
        implements MediaBrowsable, MediaControllerObserver.Listener {

    private static final String TAG = LogHelper.makeLogTag(MediaBrowserActivity.class);

    private MediaBrowserCompat mMediaBrowser;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        LogHelper.i(TAG, "Activity onCreate");
        MediaControllerObserver.getInstance().addListener(this);
        mMediaBrowser = new MediaBrowserCompat(this, new ComponentName(this, MusicService.class), mConnectionCallback, null);
        mMediaBrowser.connect();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        LogHelper.i(TAG, "Activity onDestroy");
        MediaControllerObserver.getInstance().removeListener(this);
        if (MediaControllerCompat.getMediaController(this) != null) {
            MediaControllerObserver.unregister(MediaControllerCompat.getMediaController(this));
        }
        mMediaBrowser.disconnect();
    }

    @Override
    final public void sendCustomAction(String action, Bundle extras, MediaBrowserCompat.CustomActionCallback callback) {
        if (mMediaBrowser != null && mMediaBrowser.isConnected()) {
            mMediaBrowser.sendCustomAction(action, extras, callback);
        }
    }

    @Override
    final public MediaBrowserCompat getMediaBrowser() {
        return mMediaBrowser;
    }

    @Override
    final public void onPlaybackStateChanged(@NonNull PlaybackStateCompat state) {
    }

    @Override
    final public void onMetadataChanged(MediaMetadataCompat metadata) {
        LogHelper.d(TAG, "onMetadataChanged");
        DrawerHelper.INSTANCE.updateHeader(this, metadata, mAccountHeader);
    }

    @Override
    public void onMediaControllerConnected() {
        MediaControllerCompat controller = MediaControllerCompat.getMediaController(this);
        LogHelper.d(TAG, "onMediaControllerConnected, mediaController==null? ", controller == null);
        if (controller != null) {
            onMetadataChanged(controller.getMetadata());
            onPlaybackStateChanged(controller.getPlaybackState());
        }
    }

    private final MediaBrowserCompat.ConnectionCallback mConnectionCallback =
            new MediaBrowserCompat.ConnectionCallback() {
                @Override
                public void onConnected() {
                    LogHelper.i(TAG, "onMediaControllerConnected");
                    try {
                        LogHelper.i(TAG, "connectToSession");
                        MediaControllerCompat mediaController = new MediaControllerCompat(MediaBrowserActivity.this, mMediaBrowser.getSessionToken());
                        MediaControllerCompat.setMediaController(MediaBrowserActivity.this, mediaController);
                        MediaControllerObserver.register(mediaController);
                        MediaControllerObserver.getInstance().notifyConnected();
                    } catch (RemoteException e) {
                        LogHelper.e(TAG, e, "could not connect media controller");
                    }
                }
            };

}

File: app/src/main/java/com/sjn/stamp/ui/activity/MediaBrowserActivity.kt
package com.sjn.stamp.ui.activity

import android.content.ComponentName
import android.os.Bundle
import android.os.RemoteException
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import com.sjn.stamp.MusicService
import com.sjn.stamp.ui.MediaBrowsable
import com.sjn.stamp.ui.observer.MediaBrowserObserver
import com.sjn.stamp.ui.observer.MediaControllerObserver
import com.sjn.stamp.utils.LogHelper

abstract class MediaBrowserActivity : DrawerActivity(), MediaBrowsable, MediaControllerObserver.Listener {

    override var mediaBrowser: MediaBrowserCompat? = null
    private val mediaController: MediaControllerCompat?
        get() = MediaControllerCompat.getMediaController(this)

    private val connectionCallback = object : MediaBrowserCompat.ConnectionCallback() {
        override fun onConnected() {
            try {
                LogHelper.i(TAG, "onMediaControllerConnected")
                MediaControllerCompat.setMediaController(this@MediaBrowserActivity, MediaControllerCompat(this@MediaBrowserActivity, mediaBrowser!!.sessionToken))
                MediaBrowserObserver.notifyConnected()
            } catch (e: RemoteException) {
                LogHelper.e(TAG, e, "could not connect media controller")
            }

        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        LogHelper.i(TAG, "Activity onCreate")
        MediaControllerObserver.addListener(this)
        mediaBrowser = MediaBrowserCompat(this, ComponentName(this, MusicService::class.java), connectionCallback, null)
        mediaBrowser?.connect()
    }

    override fun onDestroy() {
        super.onDestroy()
        LogHelper.i(TAG, "Activity onDestroy")
        MediaControllerObserver.removeListener(this)
        mediaBrowser?.disconnect()
    }

    override fun sendCustomAction(action: String, extras: Bundle?, callback: MediaBrowserCompat.CustomActionCallback?) {
        if (mediaBrowser?.isConnected == true) {
            mediaBrowser?.sendCustomAction(action, extras, callback)
        }
    }

    override fun onPlaybackStateChanged(state: PlaybackStateCompat?) {}

    override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
        LogHelper.d(TAG, "onMetadataChanged")
        drawer?.updateHeader(metadata)
    }

    override fun onMediaControllerConnected() {
        LogHelper.d(TAG, "onMediaControllerConnected, mediaController==null? ", mediaController == null)
        mediaController?.let {
            onMetadataChanged(it.metadata)
            onPlaybackStateChanged(it.playbackState)
        }
    }

    companion object {
        private val TAG = LogHelper.makeLogTag(MediaBrowserActivity::class.java)
    }

}
